Ask for PIN in Husqvarna Automower BLE integration

alistair23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

The integration now requires the AutoMower PIN when being setup. This ensure Home Assistant can communicate with more models of Mowers and with higher security levels.


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

All Automowers are setup with a 4 digit PIN. Depending the the model of the Automower and the security level set on the device by the user the PIN is required at boot and when performing certain operations.

The current Home Assistant integration doesn't require a PIN. It seems like higher security levels or certain models always require a PIN though. So the integration currently doesn't work for all models or configurations.

As such, let's request a PIN from users when setting up the integration so that we can use that for communicating with the mower. This should make the integration more robust for a range of different models and security levels as we can send the PIN as part of the BLE setup.

Fixes: https://github.com/home-assistant/core/issues/131321


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #131321
- This PR is related to issue: #131321
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alistair23 committed:
components/husqvarna_automower_ble: Support PIN

All Automowers are setup with a 4 digit PIN. Depending the the model of
the Automower and the security level set on the device by the user the
PIN is required at boot and when performing certain operations.

The current Home Assistant integration doesn't require a PIN for certain
models. It seems like higher security levels or certain models always
require a PIN though. So the integration currently doesn't work for all
models or configurations.

As such, let's request a PIN from users when setting up the integration
so that we can use that for communicating with the mower. This should
make the integration more robust for a range of different models and
security levels as we can send the PIN as part of the BLE setup.

Fixes: https://github.com/home-assistant/core/issues/131321
Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,7 +61,28 @@ async def async_step_bluetooth(
+
+        if user_input is not None:
+            self.pin = user_input[CONF_PIN]
+            await self.async_set_unique_id(self.address, raise_on_progress=False)
+            self._abort_if_unique_id_configured()
```
We already have the unique id set

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,7 +61,28 @@ async def async_step_bluetooth(
-        return await self.async_step_confirm()
+        return await self.async_step_bluetooth_confirm()
+
+    async def async_step_bluetooth_confirm(
```
Also, Do we now have 

Discovery -> bluetooth_confirm form -> confirm form -> create_entry?
User form -> confirm form -> create_entry?

Let's try to remove that confirm now, since clicking `submit` on the page to fill in the PIN sounds like a good enough confirm to me. Same for the user flow.

joostlek reviewed (changes requested):
So why do we need this now and what happens to the old config entry data for people who already have this set up?

joostlek commented:
Oh I read the description now (bad habit lol).

So I think instead of marking this a breaking change and have people setup their device again, let's create a way to update their configuration. There are 2 ways to do this and I need some more information about the PIN to know which one makes more sense:
1. Reauth flow, where we ask the user to reauthenticate with the device, so the user can put in the PIN. Benefit of this is that we can have everyone set the PIN or the integration does not work (as in, that is not a benefit, but it does what it should without adding a lot of `if CONF_PIN in entry.data` for when you add more features in the future that require that PIN. But my question here would be, can the PIN change?
2. A repair flow, where we kindly ask the user to provide a PIN and then just keep working like before. More graceful, but harder to maintain.

joostlek commented:
Someone just told me that the PIN can be changed, so I would recommend using a reauth flow to allow the user to provide the new PIN

alistair23 commented:
I have updated this to use the reauth flow and updated the tests to ensure 100% test coverage on the new flow

tim-devel commented:
Happy to test these changes if that helps.  Spring is coming and my mower doesn't work until this change is merged. Would be great to have this merged for the April release!  Thanks @alistair23

joostlek commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -35,11 +36,12 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         ) or await get_device(address)
         if not await mower.connect(device):
-            raise ConfigEntryNotReady
+            raise ConfigEntryAuthFailed(f"Unable to connect to device {address}")
```
is an auth failure the only reason why we cant connect?

joostlek reviewed (changes requested):
There were also other open comments

alistair23 committed:
components/husqvarna_automower_ble: Don't set unique ID twice

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
components/husqvarna_automower_ble: Remove confirm form

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
components/husqvarna_automower_ble: Raise ConfigEntryNotReady on init errors

Signed-off-by: Alistair Francis <alistair@alistair23.me>

NoRi2909 commented:
BTW: you have the typo "Autmower" in the title of this PR.

joostlek commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -23,9 +23,10 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Husqvarna Autoconnect Bluetooth from a config entry."""
     address = entry.data[CONF_ADDRESS]
+    pin = int(entry.data[CONF_PIN])
```
Shouldn't we have a `if CONF_PIN not in entry.data: raise ConfigEntryAuthFailed` so it will just prompt users to add their stuff?

alistair23 replied:
I have added this, the problem is I can't figure out how to unit test the line now

joostlek replied:
Creating a mock config entry without pin and then setting it up and then observing a flow has been created. `smartthings` has this added recently, check the test_init.py

alistair23 replied:
Thanks! Back to 100% test coverage

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -116,6 +150,63 @@ async def async_step_user(
+                if not await mower.connect(device):
+                    errors["base"] = "invalid_auth"
+
+                    return self.async_show_form(
+                        step_id="reauth_confirm",
+                        data_schema=vol.Schema(
+                            {
+                                vol.Required(CONF_ADDRESS): str,
+                                vol.Required(CONF_PIN): str,
+                            },
+                        ),
+                        errors=errors,
+                    )
```
if you move the `async_set_unique_id` part in an `else` block, you can remove this form as it will fall back to the form in line 203

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -116,6 +150,63 @@ async def async_step_user(
+            except (TimeoutError, BleakError):
+                return self.async_abort(reason="cannot_connect")
+
+            await self.async_set_unique_id(self.address, raise_on_progress=False)
+            self._abort_if_unique_id_configured()
```
if it's reauth, we should abort if the existing unique id mismatches with the one configured

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -116,6 +150,63 @@ async def async_step_user(
+            self.address = user_input[CONF_ADDRESS]
+            self.pin = user_input[CONF_PIN]
+
+            pin = int(self.pin) if self.pin else None
```
why `if self.pin`?

alistair23 replied:
To avoid running `int()` on a `None`

joostlek replied:
But self.pin can't be None right?

alistair23 replied:
Ah, true. It can't now thanks to https://github.com/home-assistant/core/pull/135440#discussion_r2011712571

alistair23 committed:
components/husqvarna_automower_ble: Improve config flow

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
components/husqvarna_automower_ble: Improve reauth unique id flow

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
components/husqvarna_automower_ble: Update strings

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,7 +63,26 @@ async def async_step_bluetooth(
-        return await self.async_step_confirm()
+        return await self.async_step_bluetooth_confirm()
+
+    async def async_step_bluetooth_confirm(
```
It is a bit inconsistent, that's probably worth fixing before merging. I would appreciate help if you have the chance as I find the config stuff very confusing

alistair23 replied:
Ok! I think I have addressed this. It seems a lot cleaner now

facek1ck commented:
Is this still being worked on? Would love to see this in one of the next releases.

tim-devel commented:
Is there a way to cherry pick this change on a normal home assistant install?  My mower doesn't work without this

alistair23 committed:
husqvarna_automower_ble: config_flow: Consolidate connection logic

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Simplify the config flow

Simplify the config flow so that we follow the same method for both
Bluetooth discovery and manual setup.

This should address:
https://github.com/home-assistant/core/pull/135440#discussion_r2026933525

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Remove excess inject_bluetooth_service_info

Signed-off-by: Alistair Francis <alistair@alistair23.me>

KevinSalmon commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +63,189 @@ async def async_step_bluetooth(
-        except (BleakError, TimeoutError) as exception:
-            LOGGER.exception("Failed to connect to device: %s", exception)
-            return self.async_abort(reason="cannot_connect")
+        except (BleakError, TimeoutError):
+            LOGGER.exception("Failed to probe device ({self.address}): {exception}")
```
```suggestion
        except (BleakError, TimeoutError) as exception:
            LOGGER.exception(f"Failed to probe device ({self.address}): {exception}")
```

alistair23 committed:
husqvarna_automower_ble: config_flow: A few fixes from user testing

Signed-off-by: Alistair Francis <alistair@alistair23.me>

NoRi2909 commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,40 @@
+          "pin": "Mower PIN"
+        },
+        "data_description": {
+          "pin": "The PIN used to secure the mower"
```
"pin": "[%key:component::husqvarna_automower::config::step::user::data_description::pin%]"

NoRi2909 replied:
"pin": "[%key:component::husqvarna_automower::config::step::user::data_description::pin%]"

NoRi2909 commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,40 @@
+          "pin": "Mower PIN"
+        },
+        "data_description": {
+          "pin": "The PIN used to secure the mower"
```
"pin": "[%key:component::husqvarna_automower::config::step::user::data_description::pin%]"

NoRi2909 commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,40 @@
-        "description": "Do you want to set up {name}? Make sure the mower is in pairing mode"
+        "description": "Do you want to set up {name}? Make sure the mower is in pairing mode",
+        "data": {
+          "pin": "Mower PIN"
```
"pin": "[%key:component::husqvarna_automower::config::step::user::data::pin%]"

NoRi2909 commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,40 @@
+      },
+      "reauth_confirm": {
+        "data": {
+          "pin": "Mower PIN"
```
"pin": "[%key:component::husqvarna_automower::config::step::user::data::pin%]"

NoRi2909 reviewed (commented):
We can replace four duplicates with references.

alistair23 committed:
husqvarna_automower_ble: Use references in the manifest

Signed-off-by: Alistair Francis <alistair@alistair23.me>

jojodat commented:
Still no idea why this hasn’t been merged yet — everything’s looks ready.
Hoping @joostlek gives it a go soon.

emontnemery commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -22,10 +22,14 @@
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Husqvarna Autoconnect Bluetooth from a config entry."""
+    if CONF_PIN not in entry.data:
+        raise ConfigEntryAuthFailed
```
This means every single user will be requested to input a PIN, even if it's not required, why do we do that? Can't we detect that a PIN is needed and initiate a reauth flow only then?

alistair23 replied:
Also, we don't really have an easy way to detect if it is or isn't required. Note that everyone who sets up a mower knows the PIN, so it isn't really secret

alistair23 replied:
We don't have a way to detect a wrong PIN unfortunately. Everything is reverse engineered and detecting a wrong PIN isn't supported at the moment.

alistair23 replied:
Ok! We can. I have updated this to detect an invalid PIN from the mower

alistair23 replied:
Not that I know of

emontnemery commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -22,10 +22,14 @@
+        raise ConfigEntryAuthFailed
+
     address = entry.data[CONF_ADDRESS]
+    pin = int(entry.data[CONF_PIN])
```
Why do we convert to int, don't we store it as a number in the config entry data? If it's for the type checker, we should do a type cast:
```suggestion
    pin = cast(int, entry.data[CONF_PIN])
```

alistair23 replied:
It's stored as a string and then we cast it to an int.

Otherwise the Python checkers complain

emontnemery replied:
I don't understand this answer. Why do we store it as a string?

emontnemery replied:
First of all, you should look into updating the form schemas to use selectors instead of the built-in voluptuous validators, you then get much better control over how the form is presented to the user.
In this case, I would suggest to look into using a `TextSelector` with the mode set to `TextSelectorType.NUMBER`.

Even with a `TextSelector` in number mode, I think you need to validate the input; maybe it's enough to pass the number to your library and check if you can connect to the mower with the supplied PIN.

alistair23 replied:
I'm now using `TextSelectorType.NUMBER` and passing the number to the Mower and checking errors from the Mower. I'm not sure there is too much else I can do, as there is a large range of Mowers it isn't obvious what is and isn't a valid PIN

emontnemery replied:
Since you convert it to int here, it means you are sure the PIN is always a number. That conversion should then happen in the config flow, and we should store it as a number IMO.

emontnemery commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -22,10 +22,14 @@
     channel_id = entry.data[CONF_CLIENT_ID]
 
-    mower = Mower(channel_id, address)
+    mower = Mower(channel_id, address, pin)
```
The PIN can be changed, don't we need to detect that and start a reauth flow?

raffoul commented:
I'm expecting for a release every month. Could we have a release date please ?

emontnemery commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +63,190 @@ async def async_step_bluetooth(
+            step_id="bluetooth_confirm",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_PIN): str,
```
Don't we need to add a validator to ensure the given PIN is numerical?
Also, can we use the library here to validate the PIN?
Same comments for the user step.

alistair23 replied:
Any ideas how to do that? Just an example would be helpful

alistair23 committed:
husqvarna_automower_ble: Detect if the PIN is invalid

This includes an update to the 0.2.4 release of the automower-ble
requirement to support this new feature.

The 0.2.2 release includes the following changes:
 - Return the response status from connect()
 - Fix blocking call reported by Home Assistant
 - remove local package references
 - Don't index into manufacturer_data to avoid out of range access
 - Add Sileno City 300
The 0.2.3 release includes:
 - Report ResponseResult NOT_ALLOWED on auth errors
 - Fix the AttributeError calling get_response_result()
The 0.2.4 release includes:
 - Fixup incorrect response checks during connect()
 - Fix is_connected() when called before the first connection
 - Add keep alive support
The 0.2.5 release includes
 - Fixup invalid major/minor validation

Signed-off-by: Alistair Francis <alistair@alistair23.me>

emontnemery commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -40,8 +41,13 @@ async def async_setup_entry(hass: HomeAssistant, entry: HusqvarnaConfigEntry) ->
-            raise ConfigEntryNotReady
+        response_result = await mower.connect(device)
+        if response_result is ResponseResult.INVALID_PIN:
+            raise ConfigEntryAuthFailed(f"Unable to connect to device {address}")
```
Nice!

alistair23 committed:
husqvarna_automower_ble: Handle authentication errors

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Support numeric PINs

Signed-off-by: Alistair Francis <alistair@alistair23.me>

emontnemery commented:
There's an issue in the library, when sending the PIN, `BLEClient.connect` dosn't return a `ResponseResult`, instead it returns a `MowerState`: 

https://github.com/alistair23/AutoMower-BLE/blob/b8f0b9bedab4f009179ce2afdff09e904bafcc9c/automower_ble/protocol.py#L483

https://github.com/alistair23/AutoMower-BLE/blob/b8f0b9bedab4f009179ce2afdff09e904bafcc9c/automower_ble/protocol.py#L638-L644

This means the check in the config flow fails.

emontnemery commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +69,225 @@ async def async_step_bluetooth(
+                elif response_result is not ResponseResult.OK:
+                    errors["base"] = "cannot_connect"
+                else:
+                    data = {
+                        CONF_ADDRESS: self.address,
+                        CONF_PIN: self.pin,
+                    }
```
Now we forget the client ID, and after reauth setting up fails with:
```
2025-07-19 11:27:20.788 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Unknown Automower for husqvarna_automower_ble
Traceback (most recent call last):
  File "/home/erik/development/home-assistant_fork/homeassistant/config_entries.py", line 749, in __async_setup_with_context
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/erik/development/home-assistant_fork/homeassistant/components/husqvarna_automower_ble/__init__.py", line 34, in async_setup_entry
    channel_id = entry.data[CONF_CLIENT_ID]
                 ~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: 'client_id'
```

emontnemery commented on tests/components/husqvarna_automower_ble/conftest.py:
Please improve the tests to check the contents of config entry data (address, client_id, pin) and config entry options (nothing) in all cases where we create or modify a config entry

alistair23 committed:
husqvarna_automower_ble: Bump automower-ble to 0.2.7

The 0.2.6 release includes the following changes:
 - Fix return value of get_response_result
 - Bump Python version used in CI to match pyproject.toml
 - Fix CI branch rule

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Convert PIN to an int

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Use suggested values

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Remove finalise functions

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +69,225 @@ async def async_step_bluetooth(
+                    errors=errors,
+                )
+        except (TimeoutError, BleakError):
+            return self.async_abort(reason="cannot_connect")
```
Yeah, it does. If it's unable to connect then we abort. Do you think we should reauth instead?

alistair23 committed:
husqvarna_automower_ble: Don't forget channel_id

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 commented on homeassistant/components/husqvarna_automower_ble/manifest.json:
```diff
@@ -12,5 +12,5 @@
   "documentation": "https://www.home-assistant.io/integrations/husqvarna_automower_ble",
   "iot_class": "local_polling",
-  "requirements": ["automower-ble==0.2.1"]
+  "requirements": ["automower-ble==0.2.6"]
```
It's not really possible to just update the dependency without a few of the other changes as there is an API change.

In saying that the requirements have now been updated, so this isn't an issue any more anyway

alistair23 committed:
husqvarna_automower_ble: Improve config entry test checks

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 commented:
> Sorry for the completely unqualified comment from me as a user. But when I see the slow back and forth over details of something that already works very well for most users who would ever use it, just to tweak edge cases, I’m afraid this will only become available for all at a time when no one needs it anymore. We’re already well past the middle of the mowing season. Maybe just release a working version now and do the fine-tuning afterwards? Please think of the many users who would be more than happy just to get anything working.

I would like to point out that the Home Assistant reviewers have caught and fixed numerous issues in their reviews. Although it has been a slow process the current code is much better quality then the original code. This makes it more reliable for users and also easier to maintain and improve in the future. They have even contributed patches back to the original library to fix issues there!

They really have been very helpful and patient during this process. Reviewing code is a tireless and thankless task, but is so important

Part of the slowness is on my part as well. This is all done in my spare time, outside a full time job a young family and other important commitments. For a feature (PIN support) that I don't even need. So I have been slow to address reviews and update the code. That is sometimes the nature of getting stuff for free

High quality open source software takes time and effort to get right and implement in a scalable and maintainable way, which is why this appears to be slow.

I would note you can manually apply this PR to get the feature early. Also mowing season is only just starting here, so it depends where you are :)

alistair23 commented:
Two open questions, otherwise this seems ready to go!

alistair23 committed:
husqvarna_automower_ble: Improve config_flow connections

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: A few code cleanups

Signed-off-by: Alistair Francis <alistair@alistair23.me>

CFenner commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -46,7 +49,8 @@ class HusqvarnaAutomowerBleConfigFlow(ConfigFlow, domain=DOMAIN):
     def __init__(self) -> None:
         """Initialize the config flow."""
-        self.address: str | None
+        self.address: str | None = None
+        self.pin: int | None = None
```
Can be moved out of `init`

CFenner commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,42 @@
+          "pin": "Mower PIN"
+        },
+        "data_description": {
+          "pin": "The PIN used to secure the mower"
```
My mower does not have a pin, what should I enter here? Maybe we describe it here.

emontnemery replied:
Note: I had the same comment, but it's hidden away above https://github.com/home-assistant/core/pull/135440#discussion_r2256781239

emontnemery replied:
OK. Unless we can reliably detect if the mower needs PIN (and I think your answer suggests we can't), we should go ahead with the PR as is.

CFenner commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +64,218 @@ async def async_step_bluetooth(
+                vol.Schema(
+                    {
+                        vol.Required(CONF_ADDRESS): str,
+                        vol.Required(CONF_PIN): int,
```
Can you confirm that this is really an `int`? This way I cannot input PINs with leading zeros like `0000` and `0011`. Also I think the length should be limited. On other integrations `PIN` is used with `str`.

emontnemery replied:
The PIN is sent as a number to the mower, not as a string. I think the input in the flow should be a string though for the reason you mention, the config flow should then try to convert it to an int and fail if that's not possible.

emontnemery replied:
I modified the config flow to revert back to requesting a string from the user, but this time there's validation for non-numerical PIN.

CFenner commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +64,218 @@ async def async_step_bluetooth(
+        """Confirm reauthentication dialog."""
+        errors: dict[str, str] = {}
+
+        if user_input:
```
On other occasions you use:

```suggestion
        if user_input is not None:
```

emontnemery replied:
Fixed

CFenner commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +64,218 @@ async def async_step_bluetooth(
+                elif response_result is not ResponseResult.OK:
+                    errors["base"] = "cannot_connect"
+                else:
+                    data = {
+                        CONF_ADDRESS: self.address,
+                        CONF_CLIENT_ID: channel_id,
+                        CONF_PIN: self.pin,
+                    }
+
+                    return self.async_update_reload_and_abort(
+                        self._get_reauth_entry(), data=data
+                    )
```
simplify

```suggestion
                    return self.async_update_reload_and_abort(
                        self._get_reauth_entry(),
                        data={
                            CONF_ADDRESS: self.address,
                            CONF_CLIENT_ID: channel_id,
                            CONF_PIN: self.pin,
                        }
                    )
```

emontnemery replied:
Fixed

CFenner commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,42 @@
+          "pin": "The PIN used to secure the mower"
         }
       },
       "confirm": {
```
```suggestion
      "bluetooth_confirm": {
```

emontnemery commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -60,62 +64,218 @@ async def async_step_bluetooth(
-    async def async_step_user(
+        title = await self.probe_mower(device)
+        if title is None:
+            return self.async_abort(reason="cannot_connect")
```
I had the same comment: https://github.com/home-assistant/core/pull/135440#discussion_r2281492846

I think we don't need to push for that to be improved in this PR, because it's existing (arguably bad) behavior, not introduced by this PR.

CFenner replied:
That’s fine, as long as it is addressed.  

I now could not add my mower, as the ble detection entry vanishes after the failed attempt. Also adding it via the ble address did not work. 

I guess I would need to hammer retry to make it work? It’s a bit annoying as I always need to run into the garden to activate the connection mode.

emontnemery commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -37,12 +42,18 @@ async def async_setup_entry(hass: HomeAssistant, entry: HusqvarnaConfigEntry) ->
-        if not await mower.connect(device):
-            raise ConfigEntryNotReady
+        response_result = await mower.connect(device)
+        if response_result == ResponseResult.INVALID_PIN:
```
PR on the library implementing this suggestion: https://github.com/alistair23/AutoMower-BLE/pull/132

alistair23 committed:
husqvarna_automower_ble: Don't ask for the address on reauth

Signed-off-by: Alistair Francis <alistair@alistair23.me>

emontnemery committed:
Merge branch 'dev' into alistair/husqvarna-pin

emontnemery committed:
Fix tests

emontnemery committed:
Revert to storing PIN as a string

emontnemery committed:
Fix minor remarks

emontnemery committed:
Fix strings

emontnemery committed:
Fix missing placeholders

emontnemery committed:
Don't abort reauth when the mower can't be reached

emontnemery reviewed (approved):
I think this PR is good to go now.
Thanks a lot for the patience @alistair23 👍 

I want a second opinion before merging since I've been pushing a few commits myself.

emontnemery committed:
Fix logs

emontnemery committed:
Convert PIN to int in config flow

emontnemery committed:
Fix reauth flow title

emontnemery committed:
Show old PIN in reauth flow

emontnemery committed:
Tweak reauth flow

emontnemery committed:
Tweak reauth flow

emontnemery commented:
@CFenner I'm not sure `The PIN which has been set in the app's security settings` is necessarily an improvement, I think `The PIN used to secure the mower` is clear?

emontnemery committed:
Improve test coverage

emontnemery committed:
Fall back to BT address for mower name in flow

CFenner commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,49 @@
-      "confirm": {
-        "description": "Do you want to set up {name}? Make sure the mower is in pairing mode"
+      "bluetooth_confirm": {
+        "description": "Do you want to set up {name}?\nMake sure the mower is in pairing mode",
```
```suggestion
        "description": "Do you want to set up {name}?\nMake sure the mower is in pairing mode.",
```

CFenner reviewed (approved):
👍

emontnemery committed:
Update homeassistant/components/husqvarna_automower_ble/strings.json

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

emontnemery committed:
Reuse channel_id in reauth

emontnemery commented:
The reauth flow was not using the stored channel id, which was why the mower had to be in pairing mode when validating the PIN.
That's fixed in the latest commit, but maybe the text in the reauth flow should be adjusted.
It's now like this:
> Please confirm the PIN for {name}.\nMake sure the mower is in pairing mode and the PIN is correct.

But pairing mode is probably never needed when just confirming the PIN.
Should we just shorten like this?
> Please confirm the PIN for {name}.

emontnemery commented on homeassistant/components/husqvarna_automower_ble/strings.json:
```diff
@@ -4,18 +4,49 @@
+        }
+      },
+      "reauth_confirm": {
+        "description": "Please confirm the PIN for {name}.\nMake sure the mower is in pairing mode and the PIN is correct.",
```
```suggestion
        "description": "Please confirm the PIN for {name}.",
```

emontnemery committed:
Update homeassistant/components/husqvarna_automower_ble/strings.json
